gsoc2023: net: http: culminating pull request

Emna-Rekik opened the pull request:
This PR includes all commits for the gsoc2023 HTTP project.

Please see each commit message for details.

Emna-Rekik committed:
tests: net: http: create empty testsuites for gsoc 2023 project

This commit adds an empty test suite for http server.
No test cases have been added in this commit.
However, creating the empty test suite provides a structure
for adding tests in subsequent commits.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for parsing URL

The test verifies that the URL is correctly stored
by comparing it with the expected value.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Check server availability

This test ensures that the server is up and running as expected.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
samples: net: sockets: http_server: Implement basic functions

Implementing server start/stop functionality,
IPv4/IPv6 support, and Gzipped Compression.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

rlubos commented on subsys/net/lib/http_server/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: Apache-2.0
```
This is more of a sample than a library at the moment. 

Having a sample is fine - and we should have one, but `subsys/` is not where it belongs. And the sample should make actual use of the library we're working on.

Generally, we really need to think on how should the HTTP2 server library look like. For sure we'll need a public API (as I wrote on Discord, the public header should be defined in `include/zephyr/net/http`. You need to think on what public functions should be exposed to the application. The most basic ones that come up to me  would be:
* `http2_server_init()`
* `http2_server_start()` 
* `http2_server_stop()`

For the resource (webpage) representation in the server, you should have a look at the API added some time ago by @cfriedt (which could likely provide more details about it): https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/http/service.h#L19

Finally, the sample app should make use of the above to create a HTTP2 server. And we should cover those functionalities in tests, just as listed in the GSoC umbrella issue.

cfriedt replied:
Actually, to expand on this a bit, the http service description was specifically designed to be more-or-less independent of any actual implementation details. 

What that allows us to do, for example, is to easily (for various definitions of easy) compare two or more different http server implementations in various sample applications and testsuites. There may be an implementation-specific / conditionally compiled source file here or there, but otherwise it should work.

Ultimately, this will allow us to quantitatively compare HTTP server implementations in Zephyr and provide graphs, evidence, etc to show that e.g. Emna's server implementation outperforms CivetWeb by a factor of 2 or something along those lines, or this implementation or that implementation is able to handle N concurrent clients while using M bytes of memory, and so on.

cfriedt replied:
I'll see if I can pull the CivetWeb implementation in for reference purposes, in a way that we can easily choose this or that based on Kconfig.

cfriedt commented on subsys/net/lib/http_server/src/headers/index_html_gz.h:
```diff
@@ -0,0 +1,16 @@
+unsigned char src_index_html_gz[] = {
```
This file should be generated at build time. Please remove it with `git rm`, and then `git commit --fixup <commit>`, followed by a `git rebase -i --autosquash <prev_commit>`

cfriedt commented on subsys/net/lib/http_server/src/index_html_gz.h:
```diff
@@ -0,0 +1,16 @@
+unsigned char src_index_html_gz[] = {
```
Same here

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for parsing http2 frames

Add a test function to ensure proper parsing of HTTP/2 frames.
Fixes #59691

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
samples: net: sockets: http_server: Add more features to the sever

Refactor the code, Add h2c upgrade support,
start implementation of the HTTP/2 protocol

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
include: zephyr: net: http: Add a public header for the http2 server

Add the necessary APIs and structures for building an HTTP/2 server.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+    ctx->sockfd = socket(config->address_family, SOCK_STREAM, 0);
+    if (ctx->sockfd < 0) {
+	perror("socket");
+	exit(EXIT_FAILURE);
```
We should not use `exit()` in the library code - we don't want to break the system in case of simple errors. Just return a negative error code here (`return -errno`).

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+	struct sockaddr_in serv_addr;
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sin_family = AF_INET;
+	serv_addr.sin_addr.s_addr = INADDR_ANY;
```
There's `ip_addr` field in the config structure, we should make use of it.

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+	}
+
+	for (int i = 0; i <= ctx->num_clients; i++) {
+	    if (ctx->client_fds[i].revents & POLLIN) {
```
A small suggestion here - if you do something like
```
 if (!(ctx->client_fds[i].revents & POLLIN)) {
	continue;
}
```
You will save yourself one extra level of indentation. This will become especially useful, if we use proper (8-space tab length) everywhere.

rlubos replied:
We should also check for POLLERR, POLLHUP and react as in case of a socket error/disconnect.

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+		    int new_socket = accept_new_client(server_fd);
+
+		    for (int j = 1; j < MAX_CLIENTS; j++) {
+			if (ctx->client_fds[j].fd == 0) {
```
I believe I've already mentioned that some time ago, but 0 is still a valid socket descriptor. So in case you want to mark unused `fds` entries, you should use negaive value (most commonly used -1), not 0.

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+			    ctx->client_fds[j].fd = new_socket;
+			    ctx->client_fds[j].events = POLLIN;
+
+			    if (j > ctx->num_clients) {
+				ctx->num_clients = j;
+			    }
```
I don't think I understand this part. Shouldn't `ctx->num_client` simply be incremented by 1?

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+			http_parser_init(&parser, HTTP_REQUEST);
+			http_parser_settings_init(&parserSettings);
+			parserSettings.on_header_field = on_header_field;
+			http_parser_execute(&parser, &parserSettings, buffer,
```
Shouldn't the HTTP1 parser be executed only after we identify we deal with HTTP1 request?

Emna-Rekik replied:
I called here because I needed to see if there is an upgrade: h2c in the header field.

rlubos replied:
The thing is - you cannot feed the data to the HTTP1 parser unless you're certain it's a HTTP1 request. Otherwise, you may get unexpected results. So the order of action should be:
* Read data on a socket until you have `strlen(preface) bytes`
* If preface matches HTTP2 request - proceed as with HTTP2 request.
* Otherwise, proceed with HTTP1 request. Only now you can feed that data to the HTTP1 parser, and search for the upgrade header. If there's a header, reply and await next request (HTTP2 this time).

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+					    BUFFER_SIZE);
+
+			/* Check the client preface */
+			if (strncmp(buffer, preface, 24) != 0 &&
```
General note - please avoid so called "magic numbers" (24 in this case). It's always difficult to identify where did the value come from (and is error prone).

Better to use `strlen(preface)` in this case.

rlubos replied:
Also, please note that we may not have received the entire preface in a single `recv()` call. In case we've received only half for example, we'll start to handle the data as HTTP1, even though it may be a perfectly valid half of the HTTP2 preface.

The code dealing with STREAM sockets should always be smart enough to deal with such cases. So for example here, on a new connection we should repeat the `recv()` until we get a full preface, and only then proceed with parsing.

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+			       "peer.\n");
+			close_client_connection(ctx, i);
+		    } else {
+			buffer[valread] = '\0';
```
Why would we need to NULL terminate this buffer anyway?

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+    http_parser_settings_init(&parserSettings);
+    parserSettings.on_url = on_url;
+
+    http_parser_execute(&parser, &parserSettings, buffer, BUFFER_SIZE);
```
We should provide the actual number of bytes received to the parser, not `BUFFER_SIZE`. Also, as mentioned earlier we should not assume that we've received the full request in a single receive call.

See how the HTTP client lib deals with the parser - https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/http/http_client.c#L452, it's always a recv()/http_parser_execute() pair.

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+    unsigned int frame_count;
+    int stream_header_id = 1;
+
+    if (hasUpgradeHeader == 0) {
```
I think this conceptually belong to the HTTP1 parsing part - update header is part of HTTP1 request. IMO we only need to reply for that and take no further action (i. e. expect HTTP2 request in the next step).

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+	}
+
+	/* Read the client data */
+	valread = recv(ctx->client_fds[i].fd, buffer, BUFFER_SIZE, 0);
```
As we discussed last week - we may lose stream data here (overwrite it). If the initial `recv()` call received more data than only preface, we lose it here.

Also, think we should rethink the design here. `recv()` should only be caleld after the `poll()` call reports there is new data to read - otherwise we may block inside `handle_http2_request()`, effectively preventing the server from handling other requests.

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+	    printf("buffer %s\n", buffer);
+
+	    /* Check the client preface */
+	    if (strncmp(buffer, "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n", 24) != 0) {
```
Wasnt' this already done in `http2_server_start()`?

Emna-Rekik replied:
yes, this was for an HTTP/1 client after it was upgraded to HTTP/2

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+
+    if (valread > 24) {
+	/* Define a new buffer for the remaining data and copy it there */
+	unsigned char ubuffer[BUFFER_SIZE];
```
We should really avoid declaring large buffers on a stack. I think we should have one buffer statically allocated per client, and use only it for communication.

rlubos replied:
And actually it should be even needed to copy the buffers, just offset the orignal buffer when providing the pointer to the function (so you don't provide `buffer` directly, but `buffer + some_offset`).

rlubos commented on subsys/net/lib/http_server/src/server.c:
```diff
@@ -0,0 +1,538 @@
+
+	/* Ckeck if payload size doesn't exceed buffer size */
+	if (pos + frames[frame_count].length > buffer_len) {
+	    break;
```
We still should be prepared for this case. I. e. the actual paylaod size may exceed the buffer we're processing, so that in case of the next `recv()` call we should treat the remainig data as a payload for the previous frame, and not try to parse the new frame header.

rlubos reviewed (commented):
@Emna-Rekik Please familiarize yourself with https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style, we should follow this standard for Zephyr code. We should use tabs for indentation (configured as 8 spaces/tab), use `snake_case` for function/variable names etc.

Emna-Rekik committed:
samples: net: sockets: http_server: Add shutdown feature

Provide a controlled mechanism for shutting down the
server
Fixes #59688

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function to check server is up

This test ensures that the server is up and running as expected.
Fixes #59686

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for IPv4/IPv6

This test ensures that the server supports IPv4 and IPv6.
Fixes #59685

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for Gzipped

This test ensures that the server supports Gzipped
Compression.
Fixes #59700

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

rlubos commented on subsys/net/lib/tftp/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
```
It seems that you've accidently removed TFTP library in https://github.com/zephyrproject-rtos/zephyr/pull/59669/commits/212bf4c5e931cc396ed0fa558ecce2b86b13607e, and then recommited it in https://github.com/zephyrproject-rtos/zephyr/pull/59669/commits/a2a34522d08a0ea825606f419100775feb7b3a90. We should not introduce such changes in the history, please amend the former commit so that it does not touch anything unrelated.

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,40 @@
+	int address_family;
+};
+
+struct http2_server_ctx {
+	int sockfd;
+	struct pollfd client_fds[MAX_CLIENTS];
+	int num_clients;
+};
```
I would suggest some modification of this (eventually).
```suggestion
struct http2_server_ctx {
	int server_fd; // for accepting new client connections
	int event_fd;
	size_t num_clients;
	struct http2_client_ctx clients[NET_HTTP2_MAX_CLIENTS];
};
```

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,40 @@
+
+#include <zephyr/posix/poll.h>
+
+#define MAX_CLIENTS 10
```
Please add a `Kconfig` item for this - e.g. `CONFIG_NET_HTTP2_MAX_CLIENTS`.
```suggestion
```

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,40 @@
+	int num_clients;
+};
+
+struct http2_frame {
```
It might be better to define `struct http2_frame` elsewhere, e.g. in a private header. The telling part for me, is that `struct http2_frame` is not referenced anywhere in this header.

cfriedt commented on subsys/net/lib/http_server/src/headers/server_functions.h:
```diff
@@ -0,0 +1,31 @@
+#include <zephyr/net/http/parser.h>
+#include <zephyr/shell/shell.h>
+
+int http2_server_stop(const struct shell *shell, size_t argc, char **argv);
```
Yeah, the shell is completely unnecessary here, I think.

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for http2

This test ensures HTTP/1.1 backward compatibility
and verify HTTP 1.1 to 2.0 upgrade
Fixes #59690
Fixes #59670

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
samples: net: sockets: http_server: Add README file

The README file provides comprehensive instructions
for setting up, running, and testing the HTTP/2 server
prototype using the Zephyr RTOS.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for shutdown

This test ensures that the server can be shut down
Fixes #59688

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
samples: net: sockets: http_server: Add Kconfig file

Provide users with the ability to configure the
maximum number of clients that the server can handle

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef HTTP2_SERVER_H
+#define HTTP2_SERVER_H
+
+#include <zephyr/posix/poll.h>
```
This can be switched to `<poll.h>` as long as `CONFIG_POSIX_API` is in `prj.conf`

cfriedt commented on samples/net/sockets/http_server/src/main.c:
```diff
@@ -0,0 +1,34 @@
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/http/server.h>
+#include <zephyr/posix/netinet/in.h>
```
For apps (samples and tests), this can be `<netinet/in.h>` if you have `CONFIG_POSIX_API=y` in your `prj.conf`.

For Zephyr library code, we should use the `<zephyr/posix/...>` prefix though.

cfriedt replied:
Also, the "canonical" way that we should include headers is now in 3 blocks.

It's kind of unofficial right now, but better to be future-proof.

Each block is ordered alphabetically / lexically

```cpp
/* private headers */
#include "private_header.h"

/* standard headers (e.g. ISO C, POSIX) */
#include <errno.h>
#include <unistd.h>

/* zephyr headers */
#include <zephyr/kernel.h>
#include <zephyr/sys/bitarray.h>
```

cfriedt commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,721 @@
+#include <zephyr/logging/log.h>
+#include <zephyr/net/http/parser.h>
+#include <zephyr/net/http/server.h>
+#include <zephyr/net/socket.h>
+#include <zephyr/posix/arpa/inet.h>
+#include <zephyr/posix/fcntl.h>
+#include <zephyr/posix/netinet/in.h>
+#include <zephyr/posix/poll.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/posix/sys/eventfd.h>
```
Just remember to add `CONFIG_POSIX_API=y` to your `prj.conf`.

Also, please separate the headers into the appropriate canonical blocks.

```suggestion
#include <sys/socket.h>
#include <arpa/inet.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <poll.h>
#include <zephyr/sys/atomic.h>
#include <sys/eventfd.h>
```

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,721 @@
+				return 0;
+			}
+
+			int valread = recv(
+				ctx->client_fds[i].fd, buffer, BUFFER_SIZE, 0);
+
+			if (valread < 0) {
+				LOG_ERR("ERROR reading from socket");
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (valread == 0) {
+				LOG_INF("Connection closed by peer.\n");
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			/* Check the client preface */
+			if (strncmp(buffer, preface, strlen(preface)) != 0) {
+				LOG_INF("Client does not support HTTP/2.\n");
+				handle_http1_request(ctx, i, &valread);
+				continue;
+			}
+
+			handle_http2_request(ctx, i, &valread);
```
This comment will be a little longer, but I'll try to give a general overview of how a stateful parsing could look like.

First of all, we want to get rid of all poll()/recv() calls outside of the main server loop (so the one here). The reason for this is that we don't want to block the server from handling other requests, in case `recv()` would block inside `handle_http2_request()`. But even with non-blocking calls it doesn't work great in the current design, as in case there are no new data, you'd stop the parsing and start over from scratch (i. e. preface handling) w/o serving the request. 

### Receving new data

What I think would be needed in the first place to achieve this goal is to redesign how we recv new data, I suggest to:
* Introduce a separate input buffer for each client (so make it a part of the client context).
* Introduce a new `offset` variable to keep track of how many bytes are currently in the buffer. This would also affect how you `recv()` new data:
```c
recv(ctx->client_fds[i].fd, buffer + offset, sizeof(buffer) - offset, 0);
```
As you don't want to overwrite the data that is already in the buffer.

### Parsing the buffer

Now, `recv()` will be providing new data to the buffer, so we need `handle_http_request` to consume it. That's where the stateful handling kicks in. If you look at the existing HTTP1 parser implementation you'll see there are lots of states, the parser also works in a way that it counsumes the new data byte by byte:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/http/http_parser.c#L784
I don't think we need it to get that complex with HTTP2 since it's binary, so easier to parse.

First of all, we need to figure out what states the server can be in. The very first thing we expect from a newly connected client is a preface - therfore, it makes sense to introduce a state that represents this stage (`HTTP_PREFACE`). Once we've received the preface, we can determine whether we deal with HTTP1 or HTTP2 request. 

For HTTP1 I believe we don't need much state representation, as most of this is handled by the existing parser. `HTTP1_REQUEST` should do, but if you see a need for more states for HTTP1 parsing, don't hesitate to add more.

For HTTP2 it gets a bit more complex, as we need to implement the parser ourselves. So we need to figure out, what is the next data we expect to get, when we proceed with HTTP2 request. For me, it seems that we should expect frame header. So we can introduce `HTTP2_FRAME_HEADER` state. Once we're done with the frame header parsing, we know what type of frame we deal with. So for a header frame, we could have `HTTP2_HEADER_FRAME` in which we parse header fields one by one. Similarly, in `HTTP2_SETTINGS_FRAME` we would parse ("consume") settings, one by one (note that they all have predefined format/size). The list goes on and on, I hope that you see where it's going. If certain frame type requires more complex parsing - you can define substates as well.

Also note, that with each frame header, you know exactly how much bytes in that frame to expect. So when you're done with parsing of the current frame ("consumed" all of the frame data), you can get back to `HTTP2_FRAME_HEADER` state.

### Some pseudocode to consider

We should consider now how could this look like in practice (note that it's just a pseudo-C code, I did not compile it, it's just for an overview).

The main input function could look like:
```c
int handle_http_request(struct http2_client_ctx *ctx)
{
int handle_http_request(struct http2_client_ctx *ctx)
{
	int ret = -EINVAL;

	do {
		switch (ctx->state) {
		case HTTP_PREFACE:
			ret = handle_http_preface(ctx);
			break;
		case HTTP1_REQUEST:
			ret = handle_http1_request(ctx);
			break;
		case HTTP2_FRAME_HEADER:
			ret = handle_http2_frame_header(ctx);
			break;
		case xxx:
			etc..
		}
	} while (ret == 0 && ctx->offset > 0);

	return ret;
}
```

We need to establish some protocol for the return values. 0 could indicate successful parsing and that the function should proceed with the next state (there might still be some data in the buffer to process). `-EAGAIN` could indicate that we expect more data to come before we can proceed with parsing. Any other negative value - parsing error, we should terminate the connection.

Now for example how could `handle_preface()` look like:
```c
int handle_http_preface(struct http2_client_ctx *ctx)
{
	if (ctx->offset < strlen(preface)) {
		/* We don't have full preface yet, get more data. */
		return -EAGAIN;
	}

	if (strncmp(ctx->buffer, preface, strlen(preface)) != 0) {
		ctx->state = HTTP1_REQUEST;
	} else {
		ctx->state = HTTP2_FRAME_HEADER;
	}

	/* Now we need to "consume" the already processed data from the buffer
	 * For simplicity, I suggest to use memmove for starters, we can
	 * optimize to some other solution if needed
	 */

	/* ctx->offset will indicate remaining data in the buffer */
	ctx->offset -= strlen(preface);
	memmove(ctx->buffer, ctx->buffer + strlen(preface), ctx->offset);

	return 0;
}
```

Emna-Rekik committed:
samples: net: sockets: http_server: Add POSIX support

Compile the program using POSIX APIs and add a
makefile for Linux.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

cfriedt commented:
Say you had a uC with a single core, not even multi-threading capable. Say that the firmware on this uC had to do lots of things like read GPIO, read / write from flash memory, serve / respond to HTTP traffic, etc
```cpp


int http_client_input_work(client_context *ctx)
{  
  // how many bytes to read? 
  // depends on http parser state
  // depends also on how much data is available on your transport
  // in Linux: 
  int bytes_available_to_read;
  int ret = ioctl(ctx->fd, FIONREAD, &bytes_available_to_read);
  // in Zephyr?
  
}

int http_client_output_work(client_context *ctx)
{  
  // how many bytes to read? 
  // depends on http parser state
  
  // depends also on how much data is available on your transport
  // in Linux: 
  int bytes_available_to_write;
  int ret = ioctl(ctx->fd, FIONWRITE, &bytes_available_to_read);
  // in Zephyr?
  
  // how many bytes we can write on ctx->fd WITHOUT blocking at all
  int bytes_to_write = MIN(bytes_available_to_write, ctx->output_buffer_size);
  
  // linear buffer (array)
  int bytes_written = write(ctx->fd, ctx->output_buffer, bytes_to_write);
  // check for errors...
  __ASSERT_NO_MSG(bytes_written == bytes_to_write);
  // downside of linear buffer is that we either keep track of offset or use memmove to move data to the front of the array
  memmove( ... );
  
  // struct ring_buffer approach
  ret = write(ctx->fd, ctx->output_ring, /* MIN(bytes_to_write, ring_buffer_bytes_to_end_of_ring) */ );
  // check for errors..
  bytes_written += ret;
  bytes_to_write -= ret;
  // drain N bytes from ring buffer (essentially update internal counters)
  
  if (bytes_to_write > 0) {
    ret = write(ctx->fd, ctx->output_ring[offs], /* MIN(bytes_to_write, ring_buffer_bytes_to_end_of_ring) */ );
    // check for errors
    // drain bytes_to_write bytes from ring buffer
  }
}

int do_http_work(void)
{
  int server_fd;
  int cancel_fd;
  struct pollfd pfd[1 /* server_fd */ + 1 /* cancellation_fd */ + max_num_clients ];
  
  // condition pfd array (POLLIN, POLLOUT)
  
  int ret = poll(pfd, ARRAY_SIZE(pfd), 0);
  switch(ret) {
  case 0:
    return;
  case -1:
    // report error to main as -EINTR
    return -errno;
  default:
    break;
  }
  
  // ret contains the number of file descriptors in pfd that have events
  for(int i = 0; ret > 0; --ret, ++i) {
    struct pollfd *p = &pfd[i];

    if (p->fd == cancel_fd) {
      if (!(ctx->client_fds[i].revents & POLLIN)) {
        http_server_shutdown();
        return;
      }
    }
    
    if (p->fd in client_fds) {
      if ((ctx->client_fds[i].revents & POLLERR) == POLLERR) {
        // handle error
        close_client_connection();
      if ((ctx->client_fds[i].revents & POLLHUP) == POLLHUP) {
        // handle socket close
        close_client_connection();
      } else if ((ctx->client_fds[i].revents & POLLOUT) == POLLOUT) {
        // there is room to write data on this socket
        // even our outgoing traffic has to be stateful!
        // so sendAll() cannot work in a fully async system
        http_client_output_work(&client_context[p->fd]);
      } else if ((ctx->client_fds[i].revents & POLLIN) == POLLIN) {
        // there is room to write data on this socket
        // even our outgoing traffic has to be stateful!
        // so sendAll() cannot work in a fully async system
        http_client_input_work(&client_context[p->fd]);
      }
    }
    
    if (p->fd == server_fd) {
      // check if server *can* accept new client (check if num clients < max)
    }
  }
}

int main()
{
  do {
    // all of these functions would need to be more or less instantaneous
    do_gpio_work();
    do_flash_memory_work();
    do_http_work();
    // sleep until some event happens
  } while(1);
}
```

Emna-Rekik committed:
samples: net: sockets: http_server: Design Improvement

Improve Asynchronous Server Design

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+			}
+
+			ctx->clients[i - 2].offset += valread;
+			ctx->clients[i - 2].server_state = INITIAL_STATE;
```
Reinitializing the client state on each loop iteration doesn't seem right - it spoils the idea of stateful server. The main reason of the state variable is to remember at which phase are we in the request processing so you cannot reset the state on each loop iteration. The state should be initialized once - when you accept new incoming connection.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+}
+
+void close_client_connection(
+	struct http2_server_ctx *ctx_server, int client_index)
```
I think it'd be better to accept `struct http2_client_ctx *client` as the second parameter. Instead of doing some not-always-obvious math on the intext, the `http2_client_ctx` could simply remember which FDS slot it uses (i. e. index in the FDS table).

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+			break;
+		default:
+			LOG_ERR("Unknown state.\n");
+			close_client_connection(ctx_server, client_index);
```
I think it shouldn't be the responsibility of `handle_http_request()` or any of the subfunctions to close the connection. This function should just return an error in case of any parsing issues, and let the main loop handle this (i.e. close the connection/client context).

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+	}
+}
+
+int handle_http_request(struct http2_server_ctx *ctx_server,
```
A good habit to work out is to mark any private functions (i.e. which are not part of the public/internal API) as `static` to reduce the chance of name collisions.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+
+int handle_initial_state(struct http2_client_ctx *ctx_client)
+{
+	if (!ctx_client->is_http2) {
```
There shouldn't be a need for `is_http2` variable - whether we deal with HTTP1 or HTTP2 request should be evaluated based on the current state we're in.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+	return 0;
+}
+
+int handle_initial_state(struct http2_client_ctx *ctx_client)
```
I don't there should be a state like `INITIAL_STATE`. The states should transition from one to another smoothly based on the parsing progress - but now it seems that we have some kind of "supervisor" state that decides what should the next state  be.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+			LOG_ERR("Unknown state.\n");
+			close_client_connection(ctx_server, client_index);
+		}
+
```
You should break the loop if `ret < 0`. And handle `-EAGAIN` as a special case - it's not an error, but an indication that more data is expected.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+			ctx_client->offset);
+
+		if (ctx_client->offset == 0) {
+			ctx_client->server_state = CLOSING;
```
We can't just close the connection if there is no more data in the current iteration. As a matter of fact, here we should just switch to the next state and let the next state handler to determine what to do next. `handle_http_preface()` did its job already - parsed the PREFACE field, now we know if we should proceed with HTTP1 or HTTP2 parsing.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+			ctx_client->server_state = CLOSING;
+		} else {
+			ctx_client->server_state =
+				determine_server_state(ctx_client->buffer);
```
`determine_server_state()` will reach to the 4th byte of the buffer (frame type), but we don't know if there's enough data in the buffer (you didn't check if (offset > 4). That's why in my initial proposal I suggested to have a separate state for parsing FRAME_HEADER, in which we'd evaluate if we have enough data to parse the entire header, and then could decide what to do next.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+int handle_http2_frame_setting(struct http2_client_ctx *ctx_client)
+{
+	printf("HTTP2_FRAME_SETTINGS\n");
+	int frame_count = parse_http2_frames(
```
I think we should drop this idea of having `parse_http2_frames()` which will attempt to parse as many frames there is in the buffer, and store their contents in an enormous array of frames - it won't scale well for embedded systems.

We should be parsing frames one by one - i. e. start in a FRAME_HEADER state always, and based on the content switch to a state that will be responsible of parsing the payload of the frame. In case of SETTINGS frame, parse settings one be one, until we consume all of the data in the current frame.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+		ctx_client->buffer, ctx_client->offset, frames);
+	print_http2_frames(frames, frame_count);
+
+	if (!is_first_flag_set(ctx_client->buffer[4])) {
```
`first_flag` is not a good name - this flag was given a very specific name in the RFC, ACK. Also, we need to keep in mind, that different frame types define different flags, so we need to consider this to create an unique name (for example the same bit will be occupied by END_STREAM flag in the headers frame). Therefore for this particular case I think `settings_ack_flag` could be a good choice.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+	print_http2_frames(frames, frame_count);
+
+	if (!is_first_flag_set(ctx_client->buffer[4])) {
+		ssize_t ret = sendall(ctx_client->client_fd, settings_frame,
```
According to the RFC:
> A SETTINGS frame MUST be sent by both endpoints at the start of a
   connection and MAY be sent at any other time by either endpoint over
   the lifetime of the connection.

This means, that we shouldn't be sending our own SETTINGS frame as a response to the clients SETTINGS, but rather proactively once the connection is established. Here, we should send SETTINGS with ACK, to acknowledge client's frame.

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,889 @@
+	return frame_count;
+}
+
+bool is_first_flag_set(unsigned char flags) { return (flags & 0x01) != 0; }
```
Also please note, that generally it'd be recommended to create defines/enums for all constants specified in the spec (flags, frame types, setting types) and use them in the code, instead of using magic numbers in various places.

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for frame type

Add test function to verify HTTP/2 frame types.
Fixes #59696

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

rlubos commented on samples/net/sockets/http_server/src/headers/config.h:
```diff
@@ -0,0 +1,47 @@
+	unsigned char *payload;
+};
+
+enum http_frame_type {
```
I'd call this `http2_frame_type` - the type is HTTP2 specific.

rlubos commented on samples/net/sockets/http_server/src/headers/config.h:
```diff
@@ -0,0 +1,47 @@
+};
+
+enum http_frame_type {
+	DATA_FRAME = 0x00,
```
The enums should be prefixed for clarify, I suggest `HTTP2_FRAME_DATA` etc.

rlubos commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,94 @@
+	INITIAL_STATE,
+	HTTP_PREFACE,
+	HTTP1_REQUEST,
+	HTTP2_FRAME_SETTINGS,
```
And we should ensure there is no name collisions, I suggest to suffix all of the state enums with `_STATE`.

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for streams

This function verifies stream functionality and ensures
that the server handles concurrent streams.
Fixes #59694
Fixes #59699

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,979 @@
+{
+	printf("HTTP2_FRAME_HEADERS\n");
+
+	ctx_client->streams[0].stream_state =
+		determine_stream_state(ctx_client->buffer);
```
Two issues with this approach:
 * First - we shouldn't be really using a fixed `streams` array index, as this effectively breaks dealing with multiple streams. Having a stream counter doesn't work either, as this will fail if we have multiple concurrent streams.
 * Second - I don't think there's a need for `determine_stream_state()` function. Especially, that with current use it isn't really useful (note that we check frame type in there, but we already know what kind of frame (HEADERS) since we're in respective state handlers.

For current stream identification, we should be using stream ID instead. Therefore, I suggest to introduce two helper functions: 
* `struct http2_stream_ctx *find_http2_stream_context(struct http2_client_ctx *ctx_client, uint32_t stream_id)`
* `struct http2_stream_ctx *allocate_http2_stream_context(struct http2_client_ctx *ctx_client, uint32_t stream_id)`

First would loop over clients `struct http2_stream_ctx streams` and return the stream context with a matching ID, NULL if not found.
Second, would loop over clients `struct http2_stream_ctx streams` and look for an "empty" stream context (i. e. stream in IDLE state). If not found, return NULL.

Additionally, `struct http2_stream_ctx *current_stream` inside client context could be useful, to remember which stream the current frame belongs to.

Now how I'd use those functions. That's where the aforementioned state transition function can be handy. For example, we've received frame header which indicates that the frame type is HEADERS frame, we call `enter_http2_frame_headers_state()` :

* First, we use `find_http2_stream_context()` to find if the corresponding stream already exists
* If not (NULL returned) we use `allocate_http2_stream_context()` and try to allocate new stream (remember, that only HEADERS frame can do that, i. e. spawn new streams!).
* In case no empty stream can be found, reject the stream (or for now just close the connection for simplicity).
* Initialize any helper variables that may be needed to track frame parsing progress (`need_pad_len` etc,  https://discord.com/channels/@me/1105531307448074281/1134064198775750658 to refresh your memory).
* Otherwise, mark the stream as OPEN and enter the `HTTP2_FRAME_HEADERS_STATE`
* The list can be extended, if there are additional conditions stated in the RFC (or implied by the implementation).

rlubos commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,979 @@
+	int ret = -EINVAL;
+
+	switch (ctx_client->streams[0].stream_state) {
+	case IDLE_STATE:
+		ret = handle_http2_idle_state(ctx_client);
+		break;
+	case OPEN_STATE:
+		ret = handle_http2_open_state(ctx_client);
+		break;
```
I don't really think we should need separate handlers for IDLE/OPEN stream state. First off, as soon we allocate the stream it should be marked as OPEN, so there shouldn't really be the case we handle IDLE state. We should focus here on parsing header fields, one by one, and on each header you parse you should decrement `current_frame->length` by the length you parsed. And when `current_frame->length` reaches 0, it means that we're done with the current frame and we should expect a new one (so state transition to HTTP2_FRAME_HEADER_STATE, again this may involve some action, like sending the response if we received STREAM_END).

For the headers that are essential for serving the request (like the resource URI the request refers to) you should have a variable within `struct stream_context`, so that once we're done with request parsing (STREAM_END received) you know which resource to fetch. There are probably more, this is just an example.

And finally, if it turns out, that `current_frame->length` is still >0, but there's no more data in the buffer to parse, it means that we did not receive entire HEADERS frame in a single TCP packet (which can happen with TCP!). In that case, return -EAGAIN, and let `poll()`/`recv()` to fill the buffer with new data. When we're back in `handle_http_request()` it should be clear that:
* we are in HTTP2_FRAME_HEADERS_STATE
* we know how much data is left to parse in current frame (`current_frame->length`),
so we should proceed with parsing individual headers.

rlubos reviewed (commented):
It looks better, clearly going in the right direction, but there's still some work upon us. I've left a few (but rather massive) comments with design suggestions, I didn't cover everything but I didn't want to overwhelm you with the feedback.

cfriedt commented on tests/net/lib/http_server/prototype/src/main.c:
```diff
@@ -0,0 +1,539 @@
+				       0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+				       0x00, 0x00, 0x00, 0x00, 0x00};
+
+ZTEST_SUITE(server_function_tests, NULL, NULL, NULL, NULL, NULL);
```
I would suggest putting this at the bottom of the file.

jukkar commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,979 @@
+{
+	strncpy(url_buffer, at, length);
+	url_buffer[length] = '\0';
+	printf("Requested URL: %s\n", url_buffer);
```
Why do we mix printf() and LOG_xxx() macros? Better to use LOG_xxx macros allways to get consistent output.
The printf() could be converted to LOG_DBG()

cfriedt commented on tests/net/lib/http_server/table/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+  # section.subsection
+  net.http.server.table:
```
@rlubos - Emna & I weren't really sure what the tag / name should be here. I.e. "table" is just used everywhere for lack of a better term, but this is referring to the static HPACK table.

rlubos replied:
I think a combination of `hpack` and `static_table` should be used. For the test, i"d use `hpack.static_table`. 

For the implementation, personally I'd:
 - rename `http_table.c` to `http_hpack.c`,
 - rename current APIs to use `static_table` instead of `table` (not to get confused with dynamic table which we don't implement yet)
 - `http_hpack.c` leaves a room for future extenstion with dynamic table, we should aslo conisder adding header parsing logic here  (according to https://datatracker.ietf.org/doc/html/rfc7541#section-6) as it's part of the HPACK spec.

cfriedt replied:
>  - rename current APIs to use `static_table` instead of `table` (not to get confused with dynamic table which we don't implement yet)

 Given that the static and dynamic table share the same key space, it's probably better _not_ to rename public API (i.e. functions) with `static_table`, since they would just need to be renamed again later if / when dynamic table support is added. It likely does make sense internally to differentiate between static and dynamic, of course.

https://httpwg.org/specs/rfc7541.html#index.address.space

What would also be good, @Emna-Rekik, is to create a static inline function in the header file to say whether a given key is static.

E.g.

```cpp
static inline bool http_hpack_key_is_static(int32_t key)
{
    return key >= 1 && key <= 61; /* use symbolic name */
}
```

The nice thing about eventually supporting a dynamic table (which would be trivial with a hash_map), is that there is only a simple constant offset (±62) required to convert from the common key space to the dynamic key space.

cfriedt commented on include/zephyr/net/http/table.h:
```diff
@@ -0,0 +1,107 @@
+/** @file
+ * @brief HTTP request methods
```
@Emna-Rekik - please clean up the doxygen to reflect the "table" tag / name (or whatever the HPACK library is).

cfriedt commented on include/zephyr/net/http/table.h:
```diff
@@ -0,0 +1,107 @@
+	HTTP_TABLE_WWW_AUTHENTICATE = 61,
+};
+
+int http_table_add(struct http2_server_ctx *ctx, enum http_table_key key, void *value);
+int http_table_get(struct http2_server_ctx *ctx, enum http_table_key key, void **value);
+int http_table_remove(struct http2_server_ctx *ctx, enum http_table_key key);
+int http_table_contains(struct http2_server_ctx *ctx, enum http_table_key key);
```
```suggestion
int http_hpack_table_add(struct http2_server_ctx *ctx, int32_t key, void *value);
int http_hpack_table_get(struct http2_server_ctx *ctx, int32_t key, void **value);
int http_hpack_table_remove(struct http2_server_ctx *ctx, int32_t key);
int http_hpack_table_contains(struct http2_server_ctx *ctx, int32_t key);
```

cfriedt commented on include/zephyr/net/http/table.h:
```diff
@@ -0,0 +1,107 @@
+extern "C" {
+#endif
+
+enum http_table_key {
```
Might be worth it to change `HTTP_TABLE` to `HTTP_HPACK_KEY` with
```shell
sed -i -e 's/HTTP_TABLE/HTTP_HPACK/g' <filename>
```

```suggestion
enum http_hpack_static_key {
```

cfriedt commented:
One thing I'm leaning towards is dropping the '2' from the majority of the structure names and constants.

Certainly many of them will be used to support http/1.1 and maybe even http/3.

I think adding the '2' needlessly complicates the http namespace.

cfriedt commented on subsys/net/lib/http/Kconfig:
```diff
@@ -37,6 +37,52 @@ config HTTP_SERVER
 
+#if HTTP_SERVER
+
+config NET_HTTP2_MAX_CLIENTS
```
```suggestion
config NET_HTTP_MAX_CLIENTS
```

cfriedt commented on subsys/net/lib/http/Kconfig:
```diff
@@ -37,6 +37,52 @@ config HTTP_SERVER
+	help
+	  This setting determines the maximum number of HTTP/2 clients that the server can handle at once.
+
+config NET_HTTP2_MAX_STREAMS
```
```suggestion
config NET_HTTP_MAX_STREAMS
```

cfriedt replied:
I would even go further to say that Kconfig options specific to the server should have the prefix `NET_HTTP_SERVER_`.

cfriedt commented on samples/net/sockets/http_server/Kconfig:
```diff
@@ -0,0 +1,50 @@
+	int "Max number of HTTP/2 clients"
+	default 3
+	range 1 100
+	help
+	  This setting determines the maximum number of HTTP/2 clients that the server can handle at once.
+
+config NET_HTTP2_MAX_STREAMS
+	int "Max number of HTTP/2 streams"
+	default 2
+	range 1 10
+	help
+	  This setting determines the maximum number of HTTP/2 streams for each client.
+
+config NET_HTTP2_CLIENT_BUFFER_SIZE
+	int "Client Buffer Size"
+	default 256
+	range 64 1024
+	help
+	  This setting determines the buffer size for each client.
+
+config MAX_HTTP_RESPONSE_SIZE
+	int "Maximum HTTP Response Size"
+	default 128
+	range 64 1024
+	help
+	  This setting determines the maximum size of the HTTP response that the server can handle.
+
+config MAX_HTTP_URL_LENGTH
+	int "Maximum HTTP URL Length"
+	default 64
+	range 32 256
+	help
+	  This setting determines the maximum length of the HTTP URL that the server can process.
+
+config NET_HTTP2_MAX_FRAME_SIZE
+	int "Maximum Frame Size"
+	default 2048
+	range 1024 4096
+	help
+	  This setting determines the maximum size of a frame that the server can handle.
```
```suggestion
```

cfriedt commented on subsys/net/lib/http/Kconfig:
```diff
@@ -37,6 +37,52 @@ config HTTP_SERVER
 	  HTTP server support.
 	  Note: this is a work-in-progress
 
+#if HTTP_SERVER
```
Changes to this Kconfig file should have been made as part of [this commit](https://github.com/zephyrproject-rtos/zephyr/pull/59669/commits/caaf12d464cc59c8f9c7d15ffe2f25a948838375) instead of [this commit](https://github.com/zephyrproject-rtos/zephyr/pull/59669/commits/83587a3cbca9317d955763542b0a177b5202b7bd).

Can you please fix-up and reword the commit messages?

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,94 @@
+};
+
+enum http2_server_state {
+	HTTP2_FRAME_HEADER_STATE,
+	HTTP_PREFACE_STATE,
+	HTTP1_REQUEST_STATE,
+	HTTP2_FRAME_DATA_STATE,
+	HTTP2_FRAME_HEADERS_STATE,
+	HTTP2_FRAME_SETTINGS_STATE,
+	HTTP2_FRAME_PRIORITY_STATE,
+	HTTP2_FRAME_WINDOW_UPDATE_STATE,
+	HTTP2_FRAME_CONTINUATION_STATE,
+	HTTP2_FRAME_PING_STATE,
+	HTTP2_FRAME_RST_STREAM_STATE,
+	HTTP2_FRAME_GOAWAY_STATE,
+	HTTP_DONE_STATE,
```
These should be switched from `HTTP2_` (and `HTTP1_`) to `HTTP_`. The server will need to support all protocol version states, but the protocol version should be handled in code, not as a special server state.

Also, I would highly recommend that http-server-specific enumerations and constants be namespaced - e.g. `HTTP_SERVER_FOO`.

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,94 @@
+};
+
+enum http2_streaming_state {
+	IDLE_STATE,
+	OPEN_STATE,
+	CLOSE_STATE
```
I would highly recommend that http-server-specific enumerations be namespaced - e.g. `HTTP_SERVER_FOO`.

cfriedt commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,979 @@
+/*
```
Please move this file to `subsys/net/lib/http/http_server.c` so that it can be built into Zephyr's HTTP server library and subsequently used by both samples and tests.

cfriedt commented on samples/net/sockets/http_server/src/server.c:
```diff
@@ -0,0 +1,979 @@
+	return new_socket;
+}
+
+int http2_server_start(struct http2_server_ctx *ctx)
```
so again, we are doing more than http2, so please replace `http2` with `http` everywhere, unless it's a header specifically describing http2-only protocol bits.

cfriedt commented on samples/net/sockets/http_server/src/headers/server_functions.h:
```diff
@@ -0,0 +1,54 @@
+/*
```
This likely needs to be moved along with `server.c` to `subsys/net/lib/http/`

To be clear, private API functions should most likely be declared separately from public API functions, and ideally private API functions would just be declared static in `server.c`.

Please move this file to `subsys/net/lib/http/http_server_functions.h`. If these are not supposed to be publicly available functions, the please declare them as st

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,94 @@
+#ifndef ZEPHYR_INCLUDE_NET_HTTP_SERVER_H_
+#define ZEPHYR_INCLUDE_NET_HTTP_SERVER_H_
+
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
```
```suggestion
#if !defined(__ZEPHYR__)
```

cfriedt commented on include/zephyr/net/http/server.h:
```diff
@@ -0,0 +1,94 @@
+#ifndef ZEPHYR_INCLUDE_NET_HTTP_SERVER_H_
+#define ZEPHYR_INCLUDE_NET_HTTP_SERVER_H_
+
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
+
+#include <stdint.h>
+#include <poll.h>
+
+#else
+
+#include <zephyr/posix/poll.h>
+
+#endif
```
```suggestion
#include <stdint.h>

#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
#include <poll.h>
#else
#include <zephyr/posix/poll.h>
#endif
```

cfriedt reviewed (commented):
A bunch of comments. Likely more coming.

cfriedt commented on samples/net/sockets/http_server/src/headers/config.h:
```diff
@@ -0,0 +1,47 @@
+#define MAX_HTTP_URL_LENGTH CONFIG_MAX_HTTP_URL_LENGTH
+#define MAX_FRAME_SIZE CONFIG_NET_HTTP2_MAX_FRAME_SIZE
+
+struct http2_frame {
+	uint32_t length;
+	uint8_t type;
+	uint8_t flags;
+	uint32_t stream_identifier;
+	unsigned char *payload;
+};
+
+enum http2_frame_type {
+	HTTP2_DATA_FRAME = 0x00,
+	HTTP2_HEADERS_FRAME = 0x01,
+	HTTP2_PRIORITY_FRAME = 0x02,
+	HTTP2_RST_STREAM_FRAME = 0x03,
+	HTTP2_SETTINGS_FRAME = 0x04,
+	HTTP2_PUSH_PROMISE_FRAME = 0x05,
+	HTTP2_PING_FRAME = 0x06,
+	HTTP2_GOAWAY_FRAME = 0x07,
+	HTTP2_WINDOW_UPDATE_FRAME = 0x08,
+	HTTP2_CONTINUATION_FRAME = 0x09
+};
+
+#define HTTP2_FLAG_SETTINGS_ACK 0x1
+#define HTTP2_FLAG_END_HEADERS 0x4
+#define HTTP2_FLAG_STREAM 0x1
+
+#define HTTP2_FRAME_HEADER_SIZE 9
+#define HTTP2_FRAME_LENGTH_OFFSET 0
+#define HTTP2_FRAME_TYPE_OFFSET 3
+#define HTTP2_FRAME_FLAGS_OFFSET 4
+#define HTTP2_FRAME_STREAM_ID_OFFSET 5
```
Can we move this to its own header? Maybe `include/zephyr/net/http/frame.h`?

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for hpack table

Test hpack static table library
Fixes #59693

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: Implement hpack table

Implement a library for hpack static table
Fixes #59693

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: Add HTTP server library

Add Zephyr HTTP library

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: Modify HTTP client

Replace net_http with net_http_client

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: Modify http Kconfig

Modify http Kconfig
Fixes #59684

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for header parser

Verifies that HTTP method and HTTP path are correctly parsed

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
samples: net: sockets: http_server: Add sample.yaml

Add sample.yaml file and adjust files after moving
server.c to lib

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

cfriedt commented on samples/net/sockets/http_server/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+    - server
+    - socket
+  integration_platforms:
+    - native_posix_64
```
I would remove this for now. The assumption that the app could build against native_posix was based on word-of-mouth statement from the POSIX arch maintainer that POSIX API could work on ARCH POSIX, but that is certainly not the case.

rlubos commented on subsys/net/lib/http/http_hpack.c:
```diff
@@ -0,0 +1,159 @@
+	return -ENOSYS;
+}
+
+char *header_parsing(struct http_client_ctx *ctx, int requested_data)
```
This function needs a considerable design change. We should not be "requesting" anything from the parsing function. Instead, we need to provide it with buffers for header name and value, and let it fill it out with content.

I suggest that a single function call parse a single HPACK header entry (therefore it'd make sense to call it `http_hpack_parse_header()`).

Parsing HPACK entry is a bit trickier than parsing frame header from insteance, as we don't know beforehand what length to expect, so you need to always check during the parsing whether there is enough data in the buffer to proceed (for example if you read Literal Header "Name Length" field, and it would indicate that the actual string is longer than the remaining data in the buffer, you can return `-EAGAIN` and start over from scratch on the next loop iteration (when more data arrives).

I think that for simplicity, we can assume that we only support HPACK headers of size lower than `CONFIG_NET_HTTP_CLIENT_BUFFER_SIZE` - otherwise it'll become much more complex task to provide a HPACK parser able to deal with partial (incomplete) headers. IMO, for embedded system this is a reasonable tradeoff, I doubt we'll have a need to support very long header fields any time soon (like cookies).

And once the function finally returns a success - it should be the caller who decides what to do with the new data. If it's a header of interest for us (like URL, Method etc), store it. If it's not, just ignore it and proceed to the next entry.

Emna-Rekik committed:
samples: net: sockets: http_server: Update README file

Convert README.md to README.rst and update it.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

cfriedt commented on include/zephyr/net/http/frame.h:
```diff
@@ -0,0 +1,39 @@
+#define HTTP_HPACK_METHOD 0
+#define HTTP_HPACK_PATH   1
+
+#define HTTP_HPACK_STATUS_200 0x88
```
These two defines conflict with the `enum http_hpack_static_key` entries of the same name

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+			}
+
+			client->offset += valread;
+			handle_http_request(ctx, client, i);
```
We need to verify the result of `handle_http_request()` here - in case of any parsing errors, the connection with the client should be closed. With exception to `-EAGAIN` which we agreed to be a special code to indicate that more data is expected by the parser.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+			ret = handle_http_frame_window_update(ctx_client);
+			break;
+		case HTTP_FRAME_RST_STREAM_STATE:
+			ret = handle_http_frame_rst_frame(ctx_server, ctx_client, client_index);
```
Missing `break`.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+			ret = handle_http_done(ctx_server, ctx_client, client_index);
+			break;
+		default:
+			LOG_ERR("Unknown state.");
```
This looks like a serious issue for the server to enter unrecognized state - could event be an assert condition. But for now, we should at least set `ret` to some error code, and break the loop (in consequence, close the connection).

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+
+	switch (ctx_client->current_frame.type) {
+	case HTTP_HEADERS_FRAME:
+		enter_http_frame_headers_state(ctx_server, ctx_client, client_index);
```
We should take a uniform appraoch here, the current mixup of `return enter_http_xxx_state()`; and `enter_http_xxx_state(); break;` doesn't look well. I suggest to do any processing needed in current state **before** the switch instruction, and follow the `return enter_http_xxx_state()` pattern.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+		enter_http_frame_priority_state(ctx_client);
+		break;
+	default:
+		return enter_http_http_done_state(ctx_server, ctx_client, client_index);
```
This would be an exception, as unrecognized frame type is clearly a parsing error. Entering DONE state is ok, but the function should explicitly return an error code here.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+
+	int parse_result = parse_http_frame_header(ctx_client);
+
+	if (parse_result == 0) {
+		return -EAGAIN;
+	} else if (parse_result == -1) {
+		return -EINVAL;
+	}
```
`parse_http_frame_header()` could simply return those codes, instead of making it return some arbitrary values and translating it here. Then you can simply check if parsing was successful or not:
```c
if (parse_result < 0) {
	return parse_result;
}
```

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+int handle_http_frame_continuation(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_FRAME_CONTINUATION_STATE");
+	ctx_client->server_state = HTTP_FRAME_HEADERS_STATE;
```
This function should be prepared to parse any additional HPACK headers.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+	print_http_frames(frame);
+
+	if (!settings_ack_flag(frame->flags)) {
+		ssize_t ret =
```
My previous comment on this seems to have been lost in action (https://github.com/zephyrproject-rtos/zephyr/pull/59669#discussion_r1275834891),  but again, we should not be sending a SETTINGS frame as a response to the client's settings frame. To comply with the RFC, it should be sent as soon as connection is established (so when the preface indicates we deal with HTTP2, we should send it rightaway then).

Emna-Rekik replied:
I tried to do this, but the client was blocked. I'll look into it further.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1038 @@
+
+	print_http_frames(frame);
+
+	if (!has_upgrade_header) {
+
+		frame->stream_identifier = 1;
+		handle_http_frame_headers(ctx_client);
+		ctx_client->server_state = HTTP_FRAME_GOAWAY_STATE;
+
+		return 0;
+	}
```
I don't think I understand this part, WINDOW_UPDATE has nothing to do with headers.

Emna-Rekik replied:
This part was for h2c upgrade, it was a special case because it initially sent me a request using the HTTP/1.1 protocol, then upgraded to HTTP/2.0. I dont receive headers frame from it that's why I had to handle it request in differrent way

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for crime

This test ensures that the server supports Gzipped
Compression.
Fixes #59700

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for rest

Add test function for basic rest api

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for json

This test ensures that the server supports JSON
Format

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: Support json format

Add JSON format support

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: Add test function for tls

Establish a TLS communication between a client and
Zephyr HTTP server.

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

cfriedt commented on subsys/net/lib/http/http_hpack.c:
```diff
@@ -0,0 +1,159 @@
+
+static inline bool http_hpack_key_is_dynamic(int32_t key)
+{
+	return key > HTTP_SERVER_HPACK_SIZE;
```
`HTTP_SERVER_HPACK_SIZE` here is actually incorrect, and I believe Robert pointed that out over a week ago. I believe the check should use `HTTP_SERVER_HPACK_WWW_AUTHENTICATE`.

Please use `uint32_t` instead of `int32_t`.

cfriedt commented on subsys/net/lib/http/http_hpack.c:
```diff
@@ -0,0 +1,159 @@
+
+int http_hpack_table_contains(struct http_client_ctx *ctx, uint32_t key)
+{
+	if (!ctx || !http_hpack_key_is_static((int32_t)key)) {
```
remove the cast please.

cfriedt commented on subsys/net/lib/http/http_hpack.c:
```diff
@@ -0,0 +1,159 @@
+	return -ENOSYS;
+}
+
+char *http_hpack_parse_header(struct http_client_ctx *ctx, int requested_data)
```
This may be better to place in `http_parser.c` since it's really just parsing http requests using the "public" http_hpack API

cfriedt commented on subsys/net/lib/http/http_hpack.c:
```diff
@@ -0,0 +1,159 @@
+
+char *http_hpack_parse_header(struct http_client_ctx *ctx, int requested_data)
+{
+	size_t i = 0;
```
```suggestion
/* there may be a better way to do this... is it possible to use sizeof() with string literals? */
/* not sure if the naming here could be better */
#define _http_server_hpack_method_strlen 8
#define _http_server_hpack_pack_strlen 6

	size_t i = 0;
```

cfriedt commented on subsys/net/lib/http/http_hpack.c:
```diff
@@ -0,0 +1,159 @@
+
+			if (http_hpack_table_get(ctx, index, &value) == 0) {
+
+				if (requested_data == HTTP_SERVER_HPACK_METHOD &&
+				    strncmp((char *)value, ":method ", 8) == 0) {
+					return (char *)value + 8;
+				} else if (requested_data == HTTP_SERVER_HPACK_PATH &&
+					   strncmp((char *)value, ":path ", 6) == 0) {
+					return (char *)value + 6;
+				}
```
```suggestion
				if (requested_data == HTTP_SERVER_HPACK_METHOD) {
					return (char *)value + _http_server_hpack_method_strlen;
				} else if (requested_data == HTTP_SERVER_HPACK_PATH) {
					return (char *)value + _http_server_hpack_path_strlen;
				}
```

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+
+bool has_upgrade_header = true;
+
+#if defined(__ZEPHYR__)
+static const struct json_obj_descr arithmetic_result_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct arithmetic_result, x, JSON_TOK_NUMBER),
+	JSON_OBJ_DESCR_PRIM(struct arithmetic_result, y, JSON_TOK_NUMBER),
+	JSON_OBJ_DESCR_PRIM(struct arithmetic_result, result, JSON_TOK_NUMBER),
+};
+#endif
```
This seems to be test-specific data in the generic http server code. I don't think it belongs here.

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+#endif
+
+	/* Create a socket */
+#if !defined(__ZEPHYR__) || defined(CONFIG_POSIX_API)
```
Just to reduce clutter, rather than to have `#ifdef`s everywhere, it probably makes sense to use `CONFIG_NET_SOCKETS_POSIX_NAMES` in `prj.conf` and to use regular POSIX names in `http_server.c` and elsewhere.

Ideally, we would just use `CONFIG_POSIX_API`, but I believe that it is still incompatible with `CONFIG_ARCH_POSIX`.

Building for Linux was really only supposed to be to get a jumpstart, so I don't even think we necessarily need to support that use case anymore.

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+};
+#endif
+
+enum tls_tag {
+	/** The Certificate Authority public key */
+	CA_CERTIFICATE_TAG,
+	/** Used for both the public and private server keys */
+	SERVER_CERTIFICATE_TAG,
+	/** Used for both the public and private client keys */
+	CLIENT_CERTIFICATE_TAG,
+};
```
This should probably go into `http_service.h` and be properly namespaced - i.e.`HTTP_SERVICE_CA_CERT_TAG`..

rlubos replied:
I don't think we'll get to the mutual authentication, I think that for now it'd be enough to focus on the server certificate/priv key pair.

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+		return -errno;
+	}
+
+	if (setsockopt(ctx->server_fd, SOL_TLS, TLS_HOSTNAME, "localhost", sizeof("localhost")) <
```
This should use the "host" field specified via `HTTP_SERVICE_DEFINE()`.

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+
+		LOG_DBG("HTTP Method: %s", method_str);
+
+		if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/results", 8) == 0) {
```
Why is a test-specific resource ("/result") hard-coded in the generic http server library code?

The http_server in it's present state very much lacks a **CRITICAL** piece of functionality - i.e. "routing" a request to a handler based on the name and other application-specific properties of the resource.

Resources must be defined using `HTTP_RESOURCE_DEFINE()`. This will place all http resources into [Iterable Sections](https://docs.zephyrproject.org/latest/kernel/iterable_sections/index.html).

That is the only way that the server can reasonably expect to iterate over a collection (array) of resources that are declared in separate compilation units.

There are already test cases demonstrating how that is done.

The idea is to iterate over each resource associated with a particular service (defined by `HTTP_SERVICE_DEFINE()`). The resource "path" can then be compared with the http request path, etc...

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+			char response[] = "HTTP/1.1 200 OK\r\nContent-Type: "
+					  "text/plain\r\n\r\nAdded successfully.\n";
+			sendall(ctx_client->client_fd, response, strlen(response));
+		} else if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/", 1) == 0) {
```
Again, why is an application-specific resource ("/") hard-coded in the http_server library code?

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+		path = http_hpack_parse_header(ctx_client, HTTP_SERVER_HPACK_PATH);
+	}
+
+	if (strcmp(method, "GET") == 0 && strcmp(path, "/") == 0) {
```
again, why do we have a hard-coded application-specific path in the generic http server code? 

See "routing" comments earlier

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+
+	print_http_frames(ctx_client);
+
+	if (ctx_client->offset < frame->length) {
```
We cannot assume we'll fit entire frame payload into the buffer, this imposes a serious limitation (I remind you that the minimum frame size we have to support is 16k of bytes, and there's no way around it as it's the minimum specified by the RFC).

That's why I suggested that `http_hpack_parse_header` needs to be reworked (https://github.com/zephyrproject-rtos/zephyr/pull/59669#discussion_r1296070668). We should not be requesting the function to find a header in the buffer, instead, it should parse a single header (next in the buffer).

The function should:
 * return -EAGAIN if the header is not yet complete (for compressed headers it simple, as its a single byte only, but literal headers need to check for name length and value length fields to determine whether the header is complete)
 * if complete, return the header contents (strings) by filling user provided buffers (that would be the simplest approach). 
 * The function could return number of bytes parsed, or a negative error code in case of parsing errors. 
 
 And of course, once header has been parsed, we should "consume" the bytes from the buffer (`memmove()`) and decrease `current_frame->length` by the number of bytes consumed. Tracking `current_frame->length` allows us to detect, when there are no more headers to be expected (i. e. we've parsed the entire frame, `current_frame->length == 0`).

cfriedt replied:
> We cannot assume we'll fit entire frame payload into the buffer, this imposes a serious limitation (I remind you that the minimum frame size we have to support is 16k of bytes, and there's no way around it as it's the minimum specified by the RFC).

We might be able to get away with explicitly listing that as a deviation in e.g. docs, Kconfig. I would shy away from a runtime warning, but potentially a compile-time warning would be nice.

Nothing should prevent end users / system designers / product makers from configuring 16k on systems that have sufficient memory, but we can likely get away with far less.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+		method = http_hpack_parse_header(ctx_client, HTTP_SERVER_HPACK_METHOD);
+		path = http_hpack_parse_header(ctx_client, HTTP_SERVER_HPACK_PATH);
+	}
+
```
Relating to my previous comment about flags parsing, we should check `current_frame->flags` after we've parsed the entire frame content (so `current_frame->length == 0`). In such case:
* If **no** END_HEADERS flag is present, this means a CONTINUATION frame is expected - we should set a bool in the stream context that we expect CONTINUATION frame, and check it when we parse the next frame header - if any other frame arrives for a stream, it should be treated as an error.
* If END_STREAM flag is present, we should be sending the response. Hovewer, there's a caveat here - in case of END_STREAM and **no** END_HEADERS flags combination, we should wait for the CONTINUATION frame before sending the reply (we may not have all info needed yet, for example URI could be sent in the CONTINUATION frame). Therefore, again we should be keeping the information that client closed the stream in the stream context structure, and sending the reply **after** getting the final CONTINUATION frame with END_HEADERS flag.
* And finally if END_HEADERS is present and **no** END_STREAM is present, no action is needed yet - this means that client sent all of the headers, but will also be sending DATA frame (which should contain END_STREAM flag). Only then we should be sending reply.

cfriedt replied:
I agree - it's definitely better to parse the http request in a step that is separate from acting on the http request.

In an asynchronous design, the parsing step / state could even be revisited multiple times before moving to the next state.

rlubos commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1235 @@
+	return 0;
+}
+
+int handle_http_frame_continuation(struct http_client_ctx *ctx_client)
```
CONTINUATION frame should be handled similarily to HEADERS frame - it carries additional HPACK headers.

Emna-Rekik committed:
net: lib: http_server: add service header file

Add service header file

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
samples: net: sockets: http_server: Support Static Ressources

Support Static Ressources

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
include: zephyr: linker: add line for http resource

add this line
ITERABLE_SECTION_ROM(http_resource_desc, 4)

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: support only zephyr

Return them to their origin after we have
done linux prototye

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: add section rom file

add section rom file

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
tests: net: lib: http_server: add socket support

add socket support

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
net: lib: http_server: Add static ressource support

Add static ressource support

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

Emna-Rekik committed:
include: zephyr: net: http: Modify service header

Modify service header file

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1182 @@
+
+bool has_upgrade_header = true;
+
+static const struct json_obj_descr arithmetic_result_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct arithmetic_result, x, JSON_TOK_NUMBER),
+	JSON_OBJ_DESCR_PRIM(struct arithmetic_result, y, JSON_TOK_NUMBER),
+	JSON_OBJ_DESCR_PRIM(struct arithmetic_result, result, JSON_TOK_NUMBER),
+};
```
this seems to be very application-specific. Why is it part of the generic server?

cfriedt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1182 @@
+#endif
+};
+
+bool has_upgrade_header = true;
```
this seems to be some state that should probably be part of an http request detail, rather than a global variable.

cfriedt commented on tests/net/lib/http_server/crime/src/main.c:
```diff
@@ -0,0 +1,42 @@
+
+	zassert_equal(len, sizeof(compressed_inc_file), "Invalid compressed file size");
+
+	int result = memcmp(data, compressed_inc_file, len);
```
I think the point of the HTTP server supporting CRiMe, is that we want the server to serve the compressed resource and we want to obtain the resource via an HTTP GET request.

This test does not exercise any HTTP server code paths at all, from what I can tell.

cfriedt replied:
The way that you would get the server to serve a specific resource is first by defining the service, and then defining the resource.

cfriedt commented on tests/net/lib/http_server/json/src/main.c:
```diff
@@ -0,0 +1,149 @@
+
+static struct k_sem server_sem;
+
+static K_THREAD_STACK_DEFINE(server_stack, STACK_SIZE);
```
The HTTP Server should define its own thread stack, I would assume.

Really, the application should just be defining resources, metadata, and potentially custom handlers.

cfriedt commented on tests/net/lib/http_server/json/src/main.c:
```diff
@@ -0,0 +1,149 @@
+	r = connect(client_fd, (struct sockaddr *)&sa, sizeof(sa));
+	zassert_not_equal(r, -1, "failed to connect (%d)", errno);
+
+	char *http1_request = "GET /results HTTP/1.1\r\n\r\n";
```
Is it possible to use the existing http client library to perform a get request?

cfriedt commented on tests/net/lib/http_server/prototype/src/main.c:
```diff
@@ -0,0 +1,466 @@
+HTTP_RESOURCE_DEFINE(index_html_gz_resource, test_http_service, "/",
+		     &index_html_gz_resource_detail);
+
+static void server_thread_fn(void *arg0, void *arg1, void *arg2)
```
the server thread should really be an implementation detail in `subsys/net/.../http_server.c`

cfriedt commented on tests/net/lib/http_server/prototype/src/main.c:
```diff
@@ -0,0 +1,466 @@
+	close(server_fd);
+}
+
+int http_server_stop(struct http_server_ctx *ctx)
```
this should be a part of the http server implementation not the app?

Emna-Rekik committed:
tests: net: lib: http_server: Add test functions for prototype

Update prototype test functions and fix issues
Fixes #59685
Fixes #59686
Fixes #59688
Fixes #59690
Fixes #59670

Signed-off-by: Emna Rekik <emna.rekik007@gmail.com>

jukkar commented:
Hi all, what is the plan with this one (considering merging)? There are lot of commits (a bit too much imho), and reviewing becomes harder. Is there something major still missing from this?
Should we just remove the draft status and have a normal PR so that more people would look at it?
Also we might consider merging this even if not ready so that remaining issues could be resolved more easily. We could add to documentation that this is still very experimental. WDYT?

jukkar commented:
ping @cfriedt, any ideas / suggestions regarding my question?

cfriedt commented:
> ping @cfriedt, any ideas / suggestions regarding my question?
> 

@jukkar - are you on Discord by any chance?

jukkar commented:
> are you on Discord by any chance?

Yes, my nick is jukkar

pdgendt commented on subsys/net/lib/http/http_server.c:
```diff
@@ -0,0 +1,1205 @@
+			break;
+		}
+		/* Create a socket */
+		ctx->server_fd = socket(af, SOCK_STREAM, proto);
```
Isn't this an issue if there is more than 1 service? The `server_fd` member will have the value of the last socket, which is an issue for accepting new clients.

jukkar commented:
Moving the work to #64465, please add comments there. Closing this one.
